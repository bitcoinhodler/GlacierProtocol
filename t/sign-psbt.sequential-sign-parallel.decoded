{
  "tx": {
    "txid": "acb7a01d7523767a3f9c076b0ba40158a5fc02c3fcf23cb0ded7c33f6b848fa3",
    "hash": "acb7a01d7523767a3f9c076b0ba40158a5fc02c3fcf23cb0ded7c33f6b848fa3",
    "version": 2,
    "size": 82,
    "vsize": 82,
    "weight": 328,
    "locktime": 0,
    "vin": [
      {
        "txid": "f0ab814890a2561e1ffbbf947cb5574719a0b620d970af6396a60f209d173f12",
        "vout": 1,
        "scriptSig": {
          "asm": "",
          "hex": ""
        },
        "sequence": 4294967295
      }
    ],
    "vout": [
      {
        "value": 0.04999692,
        "n": 0,
        "scriptPubKey": {
          "asm": "0 a410c764ac8da12b578a06027f11e02f07b7e1d3",
          "hex": "0014a410c764ac8da12b578a06027f11e02f07b7e1d3",
          "address": "bcrt1q5sgvwe9v3ksjk4u2qcp87y0q9urm0cwnrkhp23",
          "type": "witness_v0_keyhash"
        }
      }
    ]
  },
  "unknown": {
  },
  "inputs": [
    {
      "witness_utxo": {
        "amount": 0.05000000,
        "scriptPubKey": {
          "asm": "0 65c863c750331cc03efb4daca217e7925cff87f5e9ca9927564234be0d6278e8",
          "hex": "002065c863c750331cc03efb4daca217e7925cff87f5e9ca9927564234be0d6278e8",
          "type": "witness_v0_scripthash",
          "address": "bcrt1qvhyx836sxvwvq0hmfkk2y9l8jfw0lpl4a89fjf6kgg6turtz0r5q2wgvlj"
        }
      },
      "partial_signatures": {
        "03d14ddcfb6817f5579695bbb3eb3e185887bf2942b031e6f716343b8fe7e9e8e2": "304402201712b767bb1b93bef79502f3ddea27fdffecc18c91c47a226fa5ad62fb97797c02206b384dade4b7f881d47aa90d5d3b91ea008a006b05594f8a2f632d50bf2e990b01"
      },
      "witness_script": {
        "asm": "2 03d14ddcfb6817f5579695bbb3eb3e185887bf2942b031e6f716343b8fe7e9e8e2 028fcd46f8614b2cbf318096968242a1e22bcfb6d8f2b6dc939c8c27c347b2937b 0315acb550120f4cdcb460d5c49080ab508ca4ccd8dedecac22feeac8cd017d4c1 022b063ee2f22f9e1982c140fe778672c683111a796dcccdbd47bb65e3d608a983 4 OP_CHECKMULTISIG",
        "hex": "522103d14ddcfb6817f5579695bbb3eb3e185887bf2942b031e6f716343b8fe7e9e8e221028fcd46f8614b2cbf318096968242a1e22bcfb6d8f2b6dc939c8c27c347b2937b210315acb550120f4cdcb460d5c49080ab508ca4ccd8dedecac22feeac8cd017d4c121022b063ee2f22f9e1982c140fe778672c683111a796dcccdbd47bb65e3d608a98354ae",
        "type": "multisig"
      }
    }
  ],
  "outputs": [
    {
    }
  ],
  "fee": 0.00000308
}

{
  "tx": {
    "txid": "acb7a01d7523767a3f9c076b0ba40158a5fc02c3fcf23cb0ded7c33f6b848fa3",
    "hash": "acb7a01d7523767a3f9c076b0ba40158a5fc02c3fcf23cb0ded7c33f6b848fa3",
    "version": 2,
    "size": 82,
    "vsize": 82,
    "weight": 328,
    "locktime": 0,
    "vin": [
      {
        "txid": "f0ab814890a2561e1ffbbf947cb5574719a0b620d970af6396a60f209d173f12",
        "vout": 1,
        "scriptSig": {
          "asm": "",
          "hex": ""
        },
        "sequence": 4294967295
      }
    ],
    "vout": [
      {
        "value": 0.04999692,
        "n": 0,
        "scriptPubKey": {
          "asm": "0 a410c764ac8da12b578a06027f11e02f07b7e1d3",
          "hex": "0014a410c764ac8da12b578a06027f11e02f07b7e1d3",
          "address": "bcrt1q5sgvwe9v3ksjk4u2qcp87y0q9urm0cwnrkhp23",
          "type": "witness_v0_keyhash"
        }
      }
    ]
  },
  "unknown": {
  },
  "inputs": [
    {
      "witness_utxo": {
        "amount": 0.05000000,
        "scriptPubKey": {
          "asm": "0 65c863c750331cc03efb4daca217e7925cff87f5e9ca9927564234be0d6278e8",
          "hex": "002065c863c750331cc03efb4daca217e7925cff87f5e9ca9927564234be0d6278e8",
          "type": "witness_v0_scripthash",
          "address": "bcrt1qvhyx836sxvwvq0hmfkk2y9l8jfw0lpl4a89fjf6kgg6turtz0r5q2wgvlj"
        }
      },
      "partial_signatures": {
        "028fcd46f8614b2cbf318096968242a1e22bcfb6d8f2b6dc939c8c27c347b2937b": "30440220793f4ca4cb476e704f9b6f4c2610b930e30d3348b48b9dfe24f01e1a72dbe3ef022030e7dadf415e0765814b7a8a8b789fa9d3dbdfe3a678e51863ecef63f88a84d801"
      },
      "witness_script": {
        "asm": "2 03d14ddcfb6817f5579695bbb3eb3e185887bf2942b031e6f716343b8fe7e9e8e2 028fcd46f8614b2cbf318096968242a1e22bcfb6d8f2b6dc939c8c27c347b2937b 0315acb550120f4cdcb460d5c49080ab508ca4ccd8dedecac22feeac8cd017d4c1 022b063ee2f22f9e1982c140fe778672c683111a796dcccdbd47bb65e3d608a983 4 OP_CHECKMULTISIG",
        "hex": "522103d14ddcfb6817f5579695bbb3eb3e185887bf2942b031e6f716343b8fe7e9e8e221028fcd46f8614b2cbf318096968242a1e22bcfb6d8f2b6dc939c8c27c347b2937b210315acb550120f4cdcb460d5c49080ab508ca4ccd8dedecac22feeac8cd017d4c121022b063ee2f22f9e1982c140fe778672c683111a796dcccdbd47bb65e3d608a98354ae",
        "type": "multisig"
      }
    }
  ],
  "outputs": [
    {
    }
  ],
  "fee": 0.00000308
}
online_regtest_wallet.py combined all PSBTs to form final transaction:
{
  "txid": "acb7a01d7523767a3f9c076b0ba40158a5fc02c3fcf23cb0ded7c33f6b848fa3",
  "hash": "ebc64caaaa4fe34a9d4016e3c10e7930f073eb08bb5f8d45d4ad5fb69692a4a6",
  "version": 2,
  "size": 370,
  "vsize": 154,
  "weight": 616,
  "locktime": 0,
  "vin": [
    {
      "txid": "f0ab814890a2561e1ffbbf947cb5574719a0b620d970af6396a60f209d173f12",
      "vout": 1,
      "scriptSig": {
        "asm": "",
        "hex": ""
      },
      "txinwitness": [
        "",
        "304402201712b767bb1b93bef79502f3ddea27fdffecc18c91c47a226fa5ad62fb97797c02206b384dade4b7f881d47aa90d5d3b91ea008a006b05594f8a2f632d50bf2e990b01",
        "30440220793f4ca4cb476e704f9b6f4c2610b930e30d3348b48b9dfe24f01e1a72dbe3ef022030e7dadf415e0765814b7a8a8b789fa9d3dbdfe3a678e51863ecef63f88a84d801",
        "522103d14ddcfb6817f5579695bbb3eb3e185887bf2942b031e6f716343b8fe7e9e8e221028fcd46f8614b2cbf318096968242a1e22bcfb6d8f2b6dc939c8c27c347b2937b210315acb550120f4cdcb460d5c49080ab508ca4ccd8dedecac22feeac8cd017d4c121022b063ee2f22f9e1982c140fe778672c683111a796dcccdbd47bb65e3d608a98354ae"
      ],
      "sequence": 4294967295
    }
  ],
  "vout": [
    {
      "value": 0.04999692,
      "n": 0,
      "scriptPubKey": {
        "asm": "0 a410c764ac8da12b578a06027f11e02f07b7e1d3",
        "hex": "0014a410c764ac8da12b578a06027f11e02f07b7e1d3",
        "address": "bcrt1q5sgvwe9v3ksjk4u2qcp87y0q9urm0cwnrkhp23",
        "type": "witness_v0_keyhash"
      }
    }
  ]
}
