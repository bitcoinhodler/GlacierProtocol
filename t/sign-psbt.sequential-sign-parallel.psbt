cHNidP8BAFICAAAAAcoVLLMRNvrEeS7CG2BGErNnlyedbQdF0+cQeDn7A0oxAQAAAAD/////AQxKTAAAAAAAFgAUpBDHZKyNoStXigYCfxHgLwe34dMAAAAAAAEBK0BLTAAAAAAAIgAgZchjx1AzHMA++02sohfnklz/h/XpypknVkI0vg1ieOgBBYtSIQPRTdz7aBf1V5aVu7PrPhhYh78pQrAx5vcWNDuP5+no4iECj81G+GFLLL8xgJaWgkKh4ivPttjyttyTnIwnw0eyk3shAxWstVASD0zctGDVxJCAq1CMpMzY3t7Kwi/urIzQF9TBIQIrBj7i8i+eGYLBQP53hnLGgxEaeW3Mzb1Hu2Xj1gipg1SuAAA=
