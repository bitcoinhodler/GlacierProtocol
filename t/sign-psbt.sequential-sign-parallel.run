#!/bin/bash
set -e

# Create two single-signature PSBTs from the same input PSBT,
# which will be combined by online_regtest_walley.py.

$GLACIERSCRIPT --regtest=$1 sign-psbt << INPUT 2>&1
y
y
y
y
y
y
../../t/sign-psbt.sequential-sign-parallel.psbt
y
1
cQCrT9Ncs9729ao7jbmAWrD9z7tF64s2yKzmD6nkiLAi9sXVZWAn
y
INPUT

# Ugly: need to delete wallet, otherwise the second signing will still have
# the first key and will create two signatures.
echo First glacierscript run complete. Deleting wallet...
bitcoin-cli -regtest -rpcport=$1 -datadir=../bitcoin-data/$1 stop
# So ugly, but without this, I get timeouts restarting bitcoind:
sleep 5
rm -rf ../bitcoin-data/$1/regtest/wallets
echo Running glacierscript again for the second signature...


$GLACIERSCRIPT --regtest=$1 sign-psbt << INPUT 2>&1
y
y
y
y
y
y
../../t/sign-psbt.sequential-sign-parallel.psbt
y
1
cP65UeSDZPiTLB6CBwasWv9oJYEjRgQXhswfwcT9HscEKDcEbgy4
y
INPUT
