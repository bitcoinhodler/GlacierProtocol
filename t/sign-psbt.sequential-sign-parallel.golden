Are you running this on a computer WITHOUT a network connection of any kind? (y/n)?Have the wireless cards in this computer been physically removed? (y/n)?Are you running on battery power? (y/n)?Are you running on an operating system booted from a USB drive? (y/n)?Is your screen hidden from view of windows, cameras, and other people? (y/n)?Are smartphones and all other nearby devices turned off and in a Faraday bag? (y/n)?Input a filename located in the current directory which contains the PSBT:
0.05000000 BTC in unspent inputs from cold storage address bcrt1qvhyx836sxvwvq0hmfkk2y9l8jfw0lpl4a89fjf6kgg6turtz0r5q2wgvlj
0.04999692 BTC going to destination address bcrt1q5sgvwe9v3ksjk4u2qcp87y0q9urm0cwnrkhp23
Fee amount: 0.00000308
Confirm? (y/n): 
How many private keys will you be signing this transaction with (at least 2 required to complete transaction)?
#: Output will be a partially signed bitcoin transaction (PSBT).
Key #1: 
Is this data correct?
*** WARNING: Incorrect data may lead to loss of funds ***

0.05000000 BTC in unspent inputs from cold storage address bcrt1qvhyx836sxvwvq0hmfkk2y9l8jfw0lpl4a89fjf6kgg6turtz0r5q2wgvlj
0.04999692 BTC going to destination address bcrt1q5sgvwe9v3ksjk4u2qcp87y0q9urm0cwnrkhp23
Fee amount: 0.00000308

Signing with private keys: 
cQCrT9Ncs9729ao7jbmAWrD9z7tF64s2yKzmD6nkiLAi9sXVZWAn


Confirm? (y/n): 
Calculating transaction...

Final fee rate: 3.756097560975609756097560976 satoshis per vbyte

Incomplete PSBT (base64):
cHNidP8BAFICAAAAARI/F50gD6aWY69w2SC2oBlHV7V8lL/7Hx5WopBIgavwAQAAAAD/////AQxKTAAAAAAAFgAUpBDHZKyNoStXigYCfxHgLwe34dMAAAAAAAEBK0BLTAAAAAAAIgAgZchjx1AzHMA++02sohfnklz/h/XpypknVkI0vg1ieOgiAgPRTdz7aBf1V5aVu7PrPhhYh78pQrAx5vcWNDuP5+no4kcwRAIgFxK3Z7sbk773lQLz3eon/f/swYyRxHoib6WtYvuXeXwCIGs4Ta3kt/iB1HqpDV07keoAigBrBVlPii9jLVC/LpkLAQEFi1IhA9FN3PtoF/VXlpW7s+s+GFiHvylCsDHm9xY0O4/n6ejiIQKPzUb4YUssvzGAlpaCQqHiK8+22PK23JOcjCfDR7KTeyEDFay1UBIPTNy0YNXEkICrUIykzNje3srCL+6sjNAX1MEhAisGPuLyL54ZgsFA/neGcsaDERp5bczNvUe7ZePWCKmDVK4iBgIrBj7i8i+eGYLBQP53hnLGgxEaeW3Mzb1Hu2Xj1gipgwSZ1wDOIgYCj81G+GFLLL8xgJaWgkKh4ivPttjyttyTnIwnw0eyk3sEvpjrOyIGAxWstVASD0zctGDVxJCAq1CMpMzY3t7Kwi/urIzQF9TBBK2FEjYiBgPRTdz7aBf1V5aVu7PrPhhYh78pQrAx5vcWNDuP5+no4gRHPm0KAAA=

Transaction fingerprint (md5):
89dd16a109fbc08be9683c91f4d1fa4f
QR code for transaction written to transaction.png
First glacierscript run complete. Deleting wallet...
Bitcoin Core stopping
Running glacierscript again for the second signature...
Are you running this on a computer WITHOUT a network connection of any kind? (y/n)?Have the wireless cards in this computer been physically removed? (y/n)?Are you running on battery power? (y/n)?Are you running on an operating system booted from a USB drive? (y/n)?Is your screen hidden from view of windows, cameras, and other people? (y/n)?Are smartphones and all other nearby devices turned off and in a Faraday bag? (y/n)?Input a filename located in the current directory which contains the PSBT:
0.05000000 BTC in unspent inputs from cold storage address bcrt1qvhyx836sxvwvq0hmfkk2y9l8jfw0lpl4a89fjf6kgg6turtz0r5q2wgvlj
0.04999692 BTC going to destination address bcrt1q5sgvwe9v3ksjk4u2qcp87y0q9urm0cwnrkhp23
Fee amount: 0.00000308
Confirm? (y/n): 
How many private keys will you be signing this transaction with (at least 2 required to complete transaction)?
#: Output will be a partially signed bitcoin transaction (PSBT).
Key #1: 
Is this data correct?
*** WARNING: Incorrect data may lead to loss of funds ***

0.05000000 BTC in unspent inputs from cold storage address bcrt1qvhyx836sxvwvq0hmfkk2y9l8jfw0lpl4a89fjf6kgg6turtz0r5q2wgvlj
0.04999692 BTC going to destination address bcrt1q5sgvwe9v3ksjk4u2qcp87y0q9urm0cwnrkhp23
Fee amount: 0.00000308

Signing with private keys: 
cP65UeSDZPiTLB6CBwasWv9oJYEjRgQXhswfwcT9HscEKDcEbgy4


Confirm? (y/n): 
Calculating transaction...

Final fee rate: 3.756097560975609756097560976 satoshis per vbyte

Incomplete PSBT (base64):
cHNidP8BAFICAAAAARI/F50gD6aWY69w2SC2oBlHV7V8lL/7Hx5WopBIgavwAQAAAAD/////AQxKTAAAAAAAFgAUpBDHZKyNoStXigYCfxHgLwe34dMAAAAAAAEBK0BLTAAAAAAAIgAgZchjx1AzHMA++02sohfnklz/h/XpypknVkI0vg1ieOgiAgKPzUb4YUssvzGAlpaCQqHiK8+22PK23JOcjCfDR7KTe0cwRAIgeT9MpMtHbnBPm29MJhC5MOMNM0i0i53+JPAeGnLb4+8CIDDn2t9BXgdlgUt6iot4n6nT29/jpnjlGGPs72P4ioTYAQEFi1IhA9FN3PtoF/VXlpW7s+s+GFiHvylCsDHm9xY0O4/n6ejiIQKPzUb4YUssvzGAlpaCQqHiK8+22PK23JOcjCfDR7KTeyEDFay1UBIPTNy0YNXEkICrUIykzNje3srCL+6sjNAX1MEhAisGPuLyL54ZgsFA/neGcsaDERp5bczNvUe7ZePWCKmDVK4iBgIrBj7i8i+eGYLBQP53hnLGgxEaeW3Mzb1Hu2Xj1gipgwSZ1wDOIgYCj81G+GFLLL8xgJaWgkKh4ivPttjyttyTnIwnw0eyk3sEvpjrOyIGAxWstVASD0zctGDVxJCAq1CMpMzY3t7Kwi/urIzQF9TBBK2FEjYiBgPRTdz7aBf1V5aVu7PrPhhYh78pQrAx5vcWNDuP5+no4gRHPm0KAAA=

Transaction fingerprint (md5):
c6c12e477c308e2fd08f01083044fe06
QR code for transaction written to transaction.png
